toc/rfc/app-runtime-interfaces/rfc-draft-user-creation-by-org-managers.md

```diff
@@ -0,0 +1,63 @@
+
+This is an issue for customers that want to onboard larger developer groups into multi-tenant Cloud Foundry foundations. All developers would have to log-in into the foundation before CF roles can be assigned - which is impractical. Customers should be able to automate the complete onboarding e.g. using CF CLI or other CF API clients like a CF terraform provider.
+
+## Proposal
+
+The proposal is to grant Org Managers the right to assign CF roles to users which are not yet known in UAA. The users are created in UAA either explicitly via an enhanced `POST /v3/users` call or implicitly when an Org Manager assigns an organization role using `POST /v3/roles`.
+
+The functionality SHALL be gated by a deployment manifest configuration flag and SHALL be disabled by default.
+
+## Required Changes by Component
+
+### Cloud Controller
+
+#### CAPI Release
+
+Introduce a configuration flag `cc.allow_user_creation_by_org_manager` which is disabled by default .
+
+#### POST /v3/users
+
+Enhance the [POST /v3/users](https://v3-apidocs.cloudfoundry.org/version/3.172.0/index.html#create-a-user) endpoint by the parameters `username` and `origin`. `guid` and `username` parameters are exclusive.
+The existing user creation by `guid` SHALL NOT be changed.
+
+Required Parameters for Create User by Name
+|Name|Type|Description|
+|---|---|---|
+|username|string|The username as (to be) registered in UAA.|
+|origin|string|The origin / identity provider for the user. `origin=uaa` is not allowed as it requires a password. |
+
+
+Permitted roles for Create User by Name
+- Admin
+- OrgManager (if configuration flag `cc.allow_user_creation_by_org_manager` is enabled)
+
+CAPI will call the [POST /Users](https://docs.cloudfoundry.org/api/uaa/version/77.14.0/index.html#create-2) endpoint of the UAA using a CAPI owned UAA client with scope `scim.write` or `uaa.admin` for creating the user if it doesn't exist yet in UAA. The user is then registered in the Cloud Controller database under the user guid returned by UAA.
+
+#### POST /v3/roles
+
+If a role is created by user name and origin (requires feature flag `set_roles_by_username` to be enabled) and the configuration flag `cc.allow_user_creation_by_org_manager` is enabled, the user with the provided name and origin SHALL be created in CF UAA if the users doesn't exist yet in UAA.
+In other words: Role assignment shall succeed even if the user is still unknown to CF UAA. The user is created implicitly similar to `POST /v3/users` by username/origin described above.
+
+### CLI
+
+`cf create-user USERNAME [--origin ORIGIN]` (user creation without specifying a password) SHOULD use the CF API `POST /v3/users` instead of calling CF UAA directly so that Org Managers can create users if configuration flag `cc.allow_user_creation_by_org_manager` is enabled.
```

Comment:
Won't this be a breaking change for CLI users who:
1. Do have permissions to create users in UAA
2. Do NOT have the `cc.allow_user_creation_by_org_manager` enabled

In that case neither the CLI nor CC will attempt to create the user in UAA.

Reply:
Good point. I shifted the CLI enhancements into a 'possible future work' section with two options that address the case "user has scim.write/uaa.admin AND `cc.allow_user_creation_by_org_manager` disabled" (more or less elegantly).
There were concerns before regarding the CLI changes so taking them out of scope for now seems the right step.

@beyhan, @torsten-sap: Any concerns from your side? Developer onboarding doesn't need `cf create-user` but `cf set-org/space-role` is sufficient which will implicitly create shadow users if flag `cc.allow_user_creation_by_org_manager` is enabled.

Reply:
@stephanme From my point of view it is fine to start with the implicit shadow user creation during the role assignment. Nevertheless, I believe that from product perspective the "create-user" command should also follow. Otherwise, it seems like a gap.

@lechnerc77 Could you please also add your point of view? Thanks!

Reply:
Good catch regarding the breaking change. I think it is fine to focus in this RFC on the API side and leave the CF clients (only CF CLI in this case) as possible future improvement. The nice side-effect here is that CF CLI `cf set-org/space-role` will just work as mentioned.

Reply:
@torsten-sap I fully agree with your statement: starting with an implicit shadow user creation is okay (and customers will already be quite happy about it), as long a the full functionality will be shipped later.